rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+---
+status: proposed
+implementation: proposed
+status_last_reviewed:
+---
+
+# Consolidate Frontend Apps
+
+## Summary
+We have too many frontend rendering applications. This increases complexity, reduces understanding and slows development. We propose consolidating our apps into a single app. This will improve standardisation, reduce maintenance costs and improve the developer experience.
+
+## Problem
+
+### Applications overview
+GOV.UK is served by multiple rendering applications (apps). Each app serves different groups of pages within the website.
+
+There are currently 8 rendering apps for GOV.UK that are supported and maintained by GOV.UK teams:
+
+1. [collections]
+2. [email-alert-frontend]
+3. [feedback]
+4. [finder-frontend]
+5. [frontend]
+6. [government-frontend]
+7. [smart-answers]
```

Comment:
Although technically a rendering app, Smart Answers works very differently to other rendering apps (it doesn't render items from the Content Store, for example), and I'm not sure it's code structure would fit into the suggested format very well.

I'm sure lots of folks are aware of this, but I just want to call this out as something that will probably need special attention.

Reply:
Oh yes, we included it for completeness. My thoughts about smart-answers are that it probably needs to be deconstructed into 3 apps. The templates for rendering, a publishing app for the content, and a gem / engine / app for the business rules / logic that can be imported into both the publishing and rendering apps.

Reply:
I think whatever the solution to `smartanswers`, it will merit its own RFC. 

For this RFC, `smartanswers` was included for completeness. Realistically I think we could get to a place where most of the rendering is done in `frontend` and `frontend` does the redirects to `smartanswers` routes removing the need for `router` and `router-api`.

Reply:
> Don't discount merging it into a single app with both rendering and an admin area, would be way simpler to build than multiple apps.

I feel like we've recently seen a bunch of work in Whitehall to get away from the publishing/rendering in a single app paradigm, how would a single admin + rendering smart answers differ from that? We have things that are intermediate to that (eg places-manager, where we have an API consumed by the frontend and the admin console for that API in one app). A Smart-Answers manager, maybe, and then smart answers would get published as normal artifacts and attached to an API object? (in the same way a Place is an item in Publisher that is published as normal, but attached to a service in Places Manager which provides the backing data)

Reply:
I'd not use Whitehall as a gold standard in this area because it spent many years being an app that wasn't rendering about 90% of the content and just 10%. I think if it had been the other way around and Whitehall was still rendering 90% of the content we'd have probably just migrated the other way to get it back to 100%. The debts of that system were always the inconsistencies and sheer complexity.

I think we shouldn't form a holy rule that publishing and rendering must be separate. Obviously we've invested a lot in having an architecture that supports separation and if we have basic content that goes into Publishing API and then a rendering app that just pulls data from content store same way as other apps then you'd be best continuing that pattern. But for something like Smart Answers where we've got complex modelling system, ability to define custom logic and a relatively small number of domain concepts, surely the overhead of having multiple apps to keep in sync would be a burden more than a benefit? Also seems unlikely that you'd want separate teams to have frontend and backend given the heavy coupling.

Obviously this is all rather academic at this point given the many years of desire for a Smart Answer management system and lack of solutions mind 😄

Reply:
Well, it wouldn't end up different teams if it were like Places (there the published item is quite light, and the backing service API management is the main thing, managed by a single team). But sure. We're probably not going to solve the problem of Smart Answers in this comment thread ;)

Reply:
A reason why I think we could break up smart-answers is because we already have something similar for simple-smart-answers. Yes, it's a much simpler wizard, but it is configured in a publishing app, but the rendering is separate. 

I think issue with regular smart-answers has always been the amount of developer toil to update the rates and content that really should be editable by content designers. So, I do think it needs to be torn up and re-thought, especially as the developer numbers aren't what they used to be.

I wouldn't ever take the conversation about smart-answers off the table. The scope of it is too large to be covered by this RFC, but it does do frontend rendering and needs to be considered with the other apps.